---
layout: post
date: 2024-06-30
title: "[프로젝트 | 똑소] 스냅샷 복원, RDS 인스턴스가 private 서브넷에 생성되어 접속 안 되는 현상"
tags: [AWS, RDS, ]
categories: [프로젝트, 똑소, ]
mermaid: true
---


> 💡 본문에 나온 설정 값들은 AWS 계정을 이관하면서 지금은 사용하지 않습니다.



### 1. 이슈: RDS의 데이터가 덮어 씌어져 있는 것을 확인

<details>
  <summary>상황요약: 24-05-29 데이터 베이스 관련 이슈 내용 요약</summary>


![0](/assets/img/2024-06-30-프로젝트--똑소-스냅샷-복원-RDS-인스턴스가-private-서브넷에-생성되어-접속-안-되는-현상.md/0.png)

1. AM 12시 쯤에 팀원의 제보로 main 페이지 데이터가 조회 되지 않는 상황을 알게 되었습니다.
2. 운영 DB의 상품 데이터가 4월 20일 이후에 오늘 데이터로 되어있는 것을 확인했습니다.

[해결 방법]

1. 데이터 복원을 위해 RDS 백업 서비스를 이용하려 했습니다.
2. 문제는 운영 중인 RDS 인스턴스에서 이전 데이터로 바로 백업하려면 “**역추적**” 옵션이 켜져 있어야 하는데 추가 비용 문제로 켜두지 않았습니다.
3. 때문에 특정시점 스냅샷으로 RDS 인스턴스를 신규로 생성하는 방법을 사용했습니다.

[현재 상황]

1. 결과적으로 현재 2개의 RDS가 올라가 있습니다.
2. 아직 스크래핑 서버는 기존 RDS를 바라보고 있습니다. (`ddokso-database-cluster-1`)
3. API 서버는 신규 생성한 RDS를 바라보고 있습니다.(`ddokso-database-cluster-2-cluster`)

문제가 발생한 이유는 아직 찾지 못했고, 기존 DB는 내일이나 모래까지 살려두고 지울 예정입니다.


⇒ 상황이 정리된 후 발생한 이유를 찾았는데, DB 권한이 있던 팀원의 실수인 것으로 확인했습니다. 
이후 권한 문제를 빡빡하게 잡기 위해 지속적으로 노력중에 있습니다.



  </details>

### 2. 결론: 데이터를 특정시점 ‘스냅샷 복원’ 기능을 사용

- 스냅샷 복원 기능은 사용중인 RDS 인스턴스를 스냅샷 시점으로 돌리는게 아닌 스냅샷으로 RDB 인스턴스를 신규 생성하는 기능이다.
<details>
  <summary>스크린샷: 스냅샷 복원을 위해 RDS 인스턴스 설정</summary>


![1](/assets/img/2024-06-30-프로젝트--똑소-스냅샷-복원-RDS-인스턴스가-private-서브넷에-생성되어-접속-안-되는-현상.md/1.png)



  </details>

### 3. 문제 발생: 생성된 RDS 인스턴스에 접속이 안 되는 증상 발생


> 💡 RDS에 접속이 안 되는 경우에 확인하는 체크 사항으로 먼저 확인한다.



#### 3.1. `nslookup` 명령으로 먼저 엔드 포인트가 살아있는지 확인한다.



{% raw %}
```bash
$ nslookup ddokso-database-cluster-1-instance-1.clffrpe4nhja.ap-northeast-2.rds.amazonaws.com
```
{% endraw %}



![2](/assets/img/2024-06-30-프로젝트--똑소-스냅샷-복원-RDS-인스턴스가-private-서브넷에-생성되어-접속-안-되는-현상.md/2.png)


⇒ 이런식으로 출력이 된다면 엔드 포인트는 살아 있다.



#### 3.2. `nc` 명령으로 엔드 포인트와 `port`가 살아 있는지 확인한다.



{% raw %}
```bash
$ nc ddokso-database-cluster-1-instance-1.clffrpe4nhja.ap-northeast-2.rds.amazonaws.com 3026
```
{% endraw %}



![3](/assets/img/2024-06-30-프로젝트--똑소-스냅샷-복원-RDS-인스턴스가-private-서브넷에-생성되어-접속-안-되는-현상.md/3.png)


⇒ 스크린샷 처럼 요청이 들어가서 password를 입력하라는 출력이 뜨면 정상적으로 열려 있는 것이다.



#### 3.3. RDS 인스턴스 ‘퍼블릭 엑세스’ 확인


![4](/assets/img/2024-06-30-프로젝트--똑소-스냅샷-복원-RDS-인스턴스가-private-서브넷에-생성되어-접속-안-되는-현상.md/4.png)



#### 3.4. RDS 인스턴스가 사용하는 보안그룹(SG) 확인


![5](/assets/img/2024-06-30-프로젝트--똑소-스냅샷-복원-RDS-인스턴스가-private-서브넷에-생성되어-접속-안-되는-현상.md/5.png)



#### 3.5. VPC 서브넷 퍼블릭 엑세스 확인

- VPC 리소스 맵에서 확인

![6](/assets/img/2024-06-30-프로젝트--똑소-스냅샷-복원-RDS-인스턴스가-private-서브넷에-생성되어-접속-안-되는-현상.md/6.png)

- VPC 라우팅 테이블에서 확인

![7](/assets/img/2024-06-30-프로젝트--똑소-스냅샷-복원-RDS-인스턴스가-private-서브넷에-생성되어-접속-안-되는-현상.md/7.png)

- VPC 라우팅 테이블 편집 모드 확인

![8](/assets/img/2024-06-30-프로젝트--똑소-스냅샷-복원-RDS-인스턴스가-private-서브넷에-생성되어-접속-안-되는-현상.md/8.png)


⇒ 일반적인 상황에서는 여기까지 확인하면 해결된다.



### 4. 위의 과정을 했음에도 해결이 되지 않는다면? 

- RDS 운영환경이 나와 같은지 확인해 보자

![9](/assets/img/2024-06-30-프로젝트--똑소-스냅샷-복원-RDS-인스턴스가-private-서브넷에-생성되어-접속-안-되는-현상.md/9.png)

1. `default` VPC가 아닌 **“사용자 정의 VPC” 사용**
2. VPC 가용영역에 n개의 가용영역 존재
3. 가용영역은 각각 2개의 서브넷 사용
	1. public 서브넷
	2. private 서브넷


### 5. 문제 해결 과정과 원인 찾기


> 📌 결론부터 말하면 VPC의 private 서브넷에 RDS 인스턴스가 생성되어 외부 접근이 안된 문제였다.


먼저 나는 **‘스냅샷 복원’** 생성 콘솔의 ‘연결’ 섹션에서 가용영역을 **ap-northeast-2c로 하고 복원을 수행했다.**
(스크린샷에서 확인 할 수 있듯, 가용영역의 어떤 서브넷에 생성할지는 선택할 수 없다.)


![10](/assets/img/2024-06-30-프로젝트--똑소-스냅샷-복원-RDS-인스턴스가-private-서브넷에-생성되어-접속-안-되는-현상.md/10.png)



#### 5.1. `Default` VPC vs 생성된 VPC 구성 차이


AWS에서 기본으로 제공하는 Default VPC의 경우 가용영역당 하나의 서브넷만 존재한다.
(가용영역: **ap-northeast-2a, ap-northeast-2b, ap-northeast-2c)**


![11](/assets/img/2024-06-30-프로젝트--똑소-스냅샷-복원-RDS-인스턴스가-private-서브넷에-생성되어-접속-안-되는-현상.md/11.png)


**그러나 VPC를 생성하면 가용영역당 2개의 서브넷이 할당된다.** 
가용영역에 할당된 2개의 서브넷중 하나는 public 나머지 하나는 private로 사용되도록 기본 설정이 잡힌다.


![12](/assets/img/2024-06-30-프로젝트--똑소-스냅샷-복원-RDS-인스턴스가-private-서브넷에-생성되어-접속-안-되는-현상.md/12.png)

- 동그란 초록색 라벨이 붙은 경우 `public` 서브넷이다.
- 네모난 파랑색 라벨이 붙은 경우 `private` 서브넷이다.

**여기서 중요한 것은** “스냅샷 복원”으로 생성된 RDS 인스턴스의 콘솔의 “연결 및 보안” 색션의 정보를 봐도 가용영역은 알 수 있지만 **가용영역의 어떤 서브넷에 생성되었는지 알기 어렵다.**


![13](/assets/img/2024-06-30-프로젝트--똑소-스냅샷-복원-RDS-인스턴스가-private-서브넷에-생성되어-접속-안-되는-현상.md/13.png)


**퍼블릭 허용하고 생성했으니 당연하게 선택한 가용영역에 public 서브넷에 RDS가 생성될 것이라고 생각했다.**


![14](/assets/img/2024-06-30-프로젝트--똑소-스냅샷-복원-RDS-인스턴스가-private-서브넷에-생성되어-접속-안-되는-현상.md/14.png)



#### 5.2. 이제 생성된 RDS의 VPC의 서브넷 구성을 보자


![15](/assets/img/2024-06-30-프로젝트--똑소-스냅샷-복원-RDS-인스턴스가-private-서브넷에-생성되어-접속-안-되는-현상.md/15.png)


`public` 서브넷은 당연하게 라우팅 테이블을 통해 네트워크 연결이 되어 있다.


![16](/assets/img/2024-06-30-프로젝트--똑소-스냅샷-복원-RDS-인스턴스가-private-서브넷에-생성되어-접속-안-되는-현상.md/16.png)


**`ddokso-vpc-igw`**는 인터넷 게이트웨이로 외부 네트워크와 연결된다. 
(정확한 비유는 아니지만 데스크탑의 네트워크 카드와 비슷한 기능을 제공합니다.)


![17](/assets/img/2024-06-30-프로젝트--똑소-스냅샷-복원-RDS-인스턴스가-private-서브넷에-생성되어-접속-안-되는-현상.md/17.png)



#### 5.3. 문제의 원인


**여기서 문제가 발생한 이유를 알게된다.** ‘스냅샷 복원’으로 생성된 RDS 인스턴스가 `private` 서브넷인 `ddokso-vpc-subnet-private3-ap-northeast-2c` 생성된 것이다.


![18](/assets/img/2024-06-30-프로젝트--똑소-스냅샷-복원-RDS-인스턴스가-private-서브넷에-생성되어-접속-안-되는-현상.md/18.png)



#### 5.4. 문제 해결 방법


여기서 새로 생성하는 방법도 있었겠지만, 다시 이런 문제가 발생하지 않는 다는 보장이 없었기 때문에 현재 RDS 인스턴스가 올라간 `private` 서브넷을 `public`으로 돌렸다.

1. `ddokso-vpc-subnet-private3-ap-northeast-2c`을 네트워크 인터페이스 게이트웨이와 연결
2. `ddokso-vpc-subnet-topublic3-ap-northeast-2c` 로 이름을 변경

![19](/assets/img/2024-06-30-프로젝트--똑소-스냅샷-복원-RDS-인스턴스가-private-서브넷에-생성되어-접속-안-되는-현상.md/19.png)



#### 5.5. 최종 ddokso-vpc의 리소스 맵 확인


![20](/assets/img/2024-06-30-프로젝트--똑소-스냅샷-복원-RDS-인스턴스가-private-서브넷에-생성되어-접속-안-되는-현상.md/20.png)


⇒ 여기까지 완료후 nc 명령이 정상 작동하는 것과, db 접속이 정상적으로 되는 것을 확인했다.



### 정리

1. ‘RDS 스냅샷 복원’ 기능을 사용하기 위해 **“사용자 정의 VPC의 특정 가용 영역”**을 지정한다.
2. ‘RDS 스냅샷 복원’으로 생성된 RDS에 접속이 되지 않는다.
3. 확인해 보니 **“사용자 정의 VPC의 특정 가용 영역”**중에 `private` 서브넷에 RDS가 생성되어 외부에서 접속 자체가 불가능 했다.

지금까지 다양한 이유로 ‘RDS 스냅샷 복원 기능’을 사용해보았습니다. 하지만 이번과 같은 이슈는 한 번도 없었습니다. 어떠한 조건 때문인지 아직은 모르지만, 같은 이슈가 있는 분이 있다면 도움이 되기를 바랍니다.


---



#### 참고자료

- [**AWS - RDS Connection timed out 연결 오류 해결**](https://green-bin.tistory.com/34)
- [**[SPRING] AWS RDS Connection timed out, unable to connect to localhost 연결 오류 해결 방법**](https://aeliketodo.tistory.com/96)
